src/coreclr/src/jit/simdashwintrinsic.cpp

```diff
@@ -178,9 +189,13 @@ GenTree* Compiler::impSimdAsHWIntrinsic(NamedIntrinsic        intrinsic,
     // We want to resolve and populate the handle cache for this type even
     // if it isn't the basis for anything carried on the node.
     baseType = getBaseTypeAndSizeOfSIMDType(clsHnd, &simdSize);
-    assert(simdSize != 0);
 
-    CORINFO_CLASS_HANDLE argClass;
+    if (simdSize == 0)
+    {
+        // We get here for a devirtualization of IEquatable`1.Equals
+        assert(!isSIMDClass(clsHnd));
```

Comment:
@AndyAyersMS, `clsHnd` points to `System.IEquatable` here but `method` points to `Vector<T>.Equals`, is there a good way to get the `clsHnd` so we can just fully intrinsify?
The diff includes a few places where this seems to "just work" but there was at least one case where the handle wasn't resolved...

For example, with [private static int ClampStart<T>(ReadOnlySpan<T> span, T trimElement) where T : IEquatable<T>](https://source.dot.net/#System.Private.CoreLib/MemoryExtensions.Trim.cs,130) we were generating:
```asm
; Assembly listing for method MemoryExtensions:ClampStart(ReadOnlySpan`1,Vector`1):int
; Emitting BLENDED_CODE for X64 CPU with AVX - Windows
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 arg0         [V00,T03] (  4,  8   )   byref  ->  rcx         ld-addr-op
;  V01 arg1         [V01,T02] (  3, 12   )   byref  ->  rsi         ld-addr-op
;  V02 loc0         [V02,T00] (  6, 18   )     int  ->  rbp        
;  V03 OutArgs      [V03    ] (  1,  1   )  lclBlk (32) [rsp+0x00]   "OutgoingArgSpace"
;  V04 tmp1         [V04,T05] (  2,  5   )   byref  ->  rdi         V06._pointer(offs=0x00) P-INDEP "field V00._pointer (fldOffset=0x0)"
;  V05 tmp2         [V05,T04] (  3,  6   )     int  ->  rbx         V06._length(offs=0x08) P-INDEP "field V00._length (fldOffset=0x8)"
;* V06 tmp3         [V06    ] (  0,  0   )  struct (16) zero-ref    "Promoted implicit byref"
;  V07 tmp4         [V07    ] (  2, 16   )  simd32  ->  [rsp+0x20]   do-not-enreg[XS] addr-exposed "by-value struct argument"
;  V08 tmp5         [V08,T01] (  2, 16   )   byref  ->  rcx         "argument with side effect"
;
; Lcl frame size = 88

G_M15759_IG01:
       push     rdi
       push     rsi
       push     rbp
       push     rbx
       sub      rsp, 88
       vzeroupper 
       mov      rsi, rdx
						;; bbWeight=1    PerfScore 5.50
G_M15759_IG02:
       mov      rdi, bword ptr [rcx]
       mov      ebx, dword ptr [rcx+8]
						;; bbWeight=1    PerfScore 4.00
G_M15759_IG03:
       xor      ebp, ebp
       test     ebx, ebx
       jle      SHORT G_M15759_IG05
						;; bbWeight=1    PerfScore 1.50
G_M15759_IG04:
       mov      rcx, rsi
       movsxd   rdx, ebp
       shl      rdx, 5
       vmovupd  ymm0, ymmword ptr[rdi+rdx]
       vmovupd  ymmword ptr[rsp+20H], ymm0
       lea      rdx, bword ptr [rsp+20H]
       call     Vector`1:Equals(Vector`1):bool:this
       test     eax, eax
       je       SHORT G_M15759_IG05
       inc      ebp
       cmp      ebp, ebx
       jl       SHORT G_M15759_IG04
						;; bbWeight=4    PerfScore 51.00
G_M15759_IG05:
       mov      eax, ebp
						;; bbWeight=1    PerfScore 0.25
G_M15759_IG06:
       vzeroupper 
       add      rsp, 88
       pop      rbx
       pop      rbp
       pop      rsi
       pop      rdi
       ret      
						;; bbWeight=1    PerfScore 4.25

; Total bytes of code 81, prolog size 11, PerfScore 74.80, (MethodHash=f3f6c270) for method MemoryExtensions:ClampStart(ReadOnlySpan`1,Vector`1):int
```

But we are now generating:
```asm
; Assembly listing for method MemoryExtensions:ClampStart(ReadOnlySpan`1,Vector`1):int
; Emitting BLENDED_CODE for X64 CPU with AVX - Windows
; optimized code
; rsp based frame
; fully interruptible
; Final local variable assignments
;
;  V00 arg0         [V00,T01] (  4,  8   )   byref  ->  rcx         ld-addr-op
;  V01 arg1         [V01,T02] (  3,  6   )   byref  ->  rdx         ld-addr-op
;  V02 loc0         [V02,T00] (  6, 18   )     int  ->  rax        
;# V03 OutArgs      [V03    ] (  1,  1   )  lclBlk ( 0) [rsp+0x00]   "OutgoingArgSpace"
;  V04 tmp1         [V04,T05] (  2, 16   )  simd32  ->  mm1         "Inlining Arg"
;  V05 tmp2         [V05,T04] (  2,  5   )   byref  ->   r8         V07._pointer(offs=0x00) P-INDEP "field V00._pointer (fldOffset=0x0)"
;  V06 tmp3         [V06,T03] (  3,  6   )     int  ->  rcx         V07._length(offs=0x08) P-INDEP "field V00._length (fldOffset=0x8)"
;* V07 tmp4         [V07    ] (  0,  0   )  struct (16) zero-ref    "Promoted implicit byref"
;  V08 cse0         [V08,T06] (  2,  5   )  simd32  ->  mm0         "CSE - aggressive"
;
; Lcl frame size = 0

G_M15759_IG01:
       vzeroupper 
						;; bbWeight=1    PerfScore 1.00
G_M15759_IG02:
       mov      r8, bword ptr [rcx]
       mov      ecx, dword ptr [rcx+8]
						;; bbWeight=1    PerfScore 4.00
G_M15759_IG03:
       xor      eax, eax
       test     ecx, ecx
       jle      SHORT G_M15759_IG05
       vmovupd  ymm0, ymmword ptr[rdx]
						;; bbWeight=1    PerfScore 5.50
G_M15759_IG04:
       movsxd   rdx, eax
       shl      rdx, 5
       vmovupd  ymm1, ymmword ptr[r8+rdx]
       vcmpps   ymm1, ymm0, ymm1, 0
       vmovmskps yrdx, ymm1
       cmp      edx, 255
       jne      SHORT G_M15759_IG05
       inc      eax
       cmp      eax, ecx
       jl       SHORT G_M15759_IG04
						;; bbWeight=4    PerfScore 52.00
G_M15759_IG05:
       vzeroupper 
       ret      
						;; bbWeight=1    PerfScore 2.00

; Total bytes of code 59, prolog size 3, PerfScore 70.70, (MethodHash=f3f6c270) for method MemoryExtensions:ClampStart(ReadOnlySpan`1,Vector`1):int
```

Reply:
Is this still an issue?

Reply:
I'll check with the next diff that I do, it may have just been an ordering issue with the checks.

Reply:
It was indeed an ordering issue and it is handled correctly now 🎉